Changes to migrate to new TUF repository

lucasmrod opened the pull request:
# Changes

- orbit >= 1.38.0, when configured to connect to https://tuf.fleetctl.com (existing fleetd deployments) will now connect to https://updates.fleetdm.com and start using the metadata in path `/opt/orbit/updates-metadata.json`.
- orbit >= 1.38.0, when configured to connect to some custom TUF (not Fleet's TUFs) will copy `/opt/orbit/tuf-metadata.json` to `/opt/orbit/updates-metadata.json` (if it doesn't exist) and start using the latter.
- fleetctl `4.63.0` will now generate artifacts using https://updates.fleetdm.com by default (or a custom TUF if `--update-url` is set) and generate two (same file) metadata files `/opt/orbit/updates-metadata.json` and the legacy one to support downgrades `/opt/orbit/tuf-metadata.json`.
- fleetctl `4.62.0` when configured to use custom TUF (not Fleet's TUF) will generate just the legacy metadata file `/opt/orbit/tuf-metadata.json`.

## User stories

See "User stories" in https://github.com/fleetdm/confidential/issues/8488.

- [x] Update `update.defaultRootMetadata` and `update.DefaultURL` when the new repository is ready.
- [X] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/Committing-Changes.md#changes-files) for more information.
- [X] Added/updated tests
- [X] Manual QA for all new/changed functionality
- For Orbit and Fleet Desktop changes:
   - [X] Orbit runs on macOS, Linux and Windows. Check if the orbit feature/bugfix should only apply to one platform (`runtime.GOOS`).
   - [X] Manual QA must be performed in the three main OSs, macOS, Windows and Linux.
   - [X] Auto-update manual QA, from released version of component to new version (see [tools/tuf/test](../tools/tuf/test/README.md)).

lucasmrod commented:
I'll convert to draft to attempt to iterate a safer migration and one that would allow rolling back to 1.35.0 if a customer/user wants to.

getvictor reviewed (approved):
Looks good.

getvictor reviewed (approved):
Looks good.

Do we need to update the TUF URL in other places in the codebase, such as tuf-status.go? Or will we do that later?

zwass commented on orbit/pkg/update/update.go:
```diff
@@ -34,9 +34,21 @@ import (
 
-	defaultURL      = "https://tuf.fleetctl.com"
-	defaultRootKeys = `{"signed":{"_type":"root","spec_version":"1.0","version":4,"expires":"2024-10-06T17:47:49Z","keys":{"0cd79ade57d278957069e03a0fca6b975b95c2895fb20bdc3075f71fc19a4474":{"keytype":"ed25519","scheme":"ed25519","keyid_hash_algorithms":["sha256","sha512"],"keyval":{"public":"4627d9071a4b4a78c5ee867ea70439583b08dbe4ff23514e3bcb0a292de9406f"}},"1a4d9beb826d1ff4e036d757cfcd6e36d0f041e ...
+// Defining these as variables so that we can overwrite them during development/testing of the migration.
+var (
```
Should these be changed back to `const` now that testing is completed?

lucasmrod replied:
I've created https://github.com/fleetdm/fleet/issues/24519 to do this before releasing, because the QA team will need some of the tooling to test this locally.

zwass commented on orbit/pkg/update/update.go:
```diff
@@ -34,9 +34,21 @@ import (
+	//
+	// These used by `fleetctl package` and by orbit in the migration to the new TUF repository.
+	//
+	// TODO(lucas): Before merging to `main` update DefaultURL to "https://updates.fleetctl.com".
+	DefaultURL       = `https://tuf.fleetctl.com`
+	MetadataFileName = "updates-metadata.json"
+	// TODO(lucas): Before merging to `main` update defaultRootMetadata to the root.json when the new repository is ready.
```
Noting these TODOs

zwass reviewed (commented):
It might be helpful to add some comments in `packaging.go` and `orbit.go` indicating when (if ever) the code added here can be removed.

Could it make sense to copy the existing metadata file to the new path for cases where the update URL is configured to something else besides Fleet's URL? I am thinking that it would be nice (related to above) if we could someday eliminate some of the migration-related code.

lucasmrod commented:
> Looks good.
> 
> Do we need to update the TUF URL in other places in the codebase, such as tuf-status.go? Or will we do that later?

Yes, later (new repository is still not "active"). We'll probably update (duplicate) tooling/CI as we will keep the old repository up for some time (We've discussed something like at least 6 months.)

lucasmrod commented:
> It might be helpful to add some comments in `packaging.go` and `orbit.go` indicating when (if ever) the code added here can be removed.

Good idea. Added.

> Could it make sense to copy the existing metadata file to the new path for cases where the update URL is configured to something else besides Fleet's URL? I am thinking that it would be nice (related to above) if we could someday eliminate some of the migration-related code.

Good idea. I've changed this so that moving forward orbit 1.38.0+ will use (both Fleet's or custom TUF) `/opt/orbit/updates-metadata.json`. (Commit [Changes to use updates-metadata.json moving forward](https://github.com/fleetdm/fleet/pull/23588/commits/2f4d073ccb51c739755c97815c86d9e100c490dd).)

zwass reviewed (dismissed):
The changes lgtm! Let's remember this needs EXTENSIVE testing before going live. If there are issues we risk having to ask users to reinstall agent packages..

lukeheath commented:
Converting PR to draft until it's ready to merge.

lucasmrod commented:
@zwass We added some changes to support the scenario where 1.38.0 needs an urgent fix (or customer wants to downgrade to 1.37.0) and endpoints cannot access https://updates.fleetdm.com.

To cover that scenario we added `HasAccessToNewTUFServer`, if it returns false then orbit 1.38.0+ will resort to communicating to the old TUF. Once `HasAccessToNewTUFServer` succeeds for the first time, then it will always communicate with the new TUF (by storing a `/opt/orbit/new-tuf-checked` file).

Please re-review this [commit](https://github.com/fleetdm/fleet/pull/23588/commits/135ed5d47e2d8c1ef159f184bd9b401914c60f05).

lukeheath commented:
I'm setting this to "Ready for review" now that it's ready. Because it's so old it would be great to merge by Friday before we report KPIs :)

zwass reviewed (commented):
IIUC the check for access to the new repo only happens while the old update URL is configured. Once the access is detected, the new update URL gets configured and the check stops happening. Is that right? If so it seems reasonable to me.

lucasmrod commented:
> IIUC the check for access to the new repo only happens while the old update URL is configured. Once the access is detected, the new update URL gets configured and the check stops happening. Is that right? If so it seems reasonable to me.

Correct.

getvictor reviewed (approved):
LGTM
